testdata/project-v3-multigroup/apis/core/v1/pod_webhook.go

```diff
@@ -0,0 +1,179 @@
+	d.decoder = decoder
+	return nil
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+//+kubebuilder:webhook:path=/validate-v1-pod,mutating=false,failurePolicy=fail,sideEffects=None,groups=core,resources=pods,verbs=create;update,versions=v1,name=vpod.kb.io,admissionReviewVersions={v1,v1beta1}
+
+type PodValidator struct {
+	client.Client
+	Log     logr.Logger
+	decoder *admission.Decoder
+}
+
+var _ admission.DecoderInjector = &PodValidator{}
+
+func (v *PodValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	r := &corev1.Pod{}
+	if err := v.decoder.Decode(req, r); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Option 1. Extract .Validate*() methods to make it resemble
+	// admission.Validator to make controller-runtime users feel at home.
+	// Initially I wanted them to be private methods (i.e. .validate*())
+	// but since .default() is not possible (see option 1. in defaulter)
+	// I went with this.
+	//
+	// Two questions that come to mind with this approach:
+	//
+	// 1. Should admission.Denied be supported?
+	// 2. Should there be something in admission.Allowed message?
+
+	var validateErr error
+	switch req.Operation {
+	case admissionv1.Create:
+		validateErr = v.ValidateCreate(r)
+	case admissionv1.Update:
+		old := &corev1.Pod{}
+		if err := v.decoder.DecodeRaw(req.OldObject, old); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		validateErr = v.ValidateUpdate(r, old)
+	case admissionv1.Delete:
+		validateErr = v.ValidateDelete(r)
+	}
+
+	if validateErr != nil {
+		return admission.Errored(http.StatusBadRequest, validateErr)
+	}
+	return admission.Allowed("")
+
+	// End of option 1.
+
+	// Option 2. Simply add TODO here and remove .validate**() methods.
+
+	// TODO(user): fill in your validation logic upon object creation.
+	return admission.Errored(http.StatusBadRequest, errors.New("not implemented"))
+
+	// End of option 2.
+}
+
+func (v *PodValidator) ValidateCreate(r *corev1.Pod) error {
+	log := v.Log.WithValues("pod", client.ObjectKeyFromObject(r))
+	log.Info("validate create")
+
+	// TODO(user): fill in your validation logic upon object creation.
+	return nil
+}
+
+func (v *PodValidator) ValidateUpdate(r, old *corev1.Pod) error {
+	log := v.Log.WithValues("pod", client.ObjectKeyFromObject(r))
+	log.Info("validate udpate")
+
+	// TODO(user): fill in your validation logic upon object update.
+	return nil
+}
+
+func (v *PodValidator) ValidateDelete(r *corev1.Pod) error {
+	log := v.Log.WithValues("pod", client.ObjectKeyFromObject(r))
+	log.Info("validate delete")
+
+	// TODO(user): fill in your validation logic upon object deletion.
+	return nil
+}
+func (v *PodValidator) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	hookServer := mgr.GetWebhookServer()
+	hookServer.Register("/validate-v1-pod", &webhook.Admission{Handler: v})
+	return nil
+}
```

Comment:
@estroz `SetupWebhookWithManager` is already there. There is also some example wiring in the `main.go`.

Reply:
Yes I saw that. I think we could come up with a convenience wrapper for a set of handlers to register them in a loop. Not necessary right now though.